homeassistant/components/template/sensor.py

```diff
@@ -147,13 +154,22 @@ def _async_create_template_tracking_entities(
         if unique_id and unique_id_prefix:
             unique_id = f"{unique_id_prefix}-{unique_id}"
 
-        sensors.append(
-            SensorTemplate(
-                hass,
-                entity_conf,
-                unique_id,
+        if entity_conf.get(CONF_RESTORE, False):
+            sensors.append(
+                SensorRestoreTemplate(
+                    hass,
+                    entity_conf,
+                    unique_id,
+                )
+            )
+        else:
+            sensors.append(
+                SensorTemplate(
+                    hass,
+                    entity_conf,
+                    unique_id,
+                )
```

Comment:
Can `CONF_RESTORE` be a parameter to `SensorTemplate` instead, or is there some benefit of splitting in two classes?

Reply:
Understood, but it makes the code much more complicated.
Instead of implementing additional classes, overload `async_internal_will_remove_from_hass` and only call `RestoreEntity().async_internal_will_remove_from_hass` if the restore flag is set.

Reply:
OK, so have if inherit from RestoreEntity and then implement async_internal_added_to_hass and async_internal_will_remove_from_hass so that the methods in RestoreEntity are only called when the entity is to be restored correct?

> Also, what happens if the user first has restore enabled, disables it, and then enables it again; will an old saved state now be restored?
Yes, at that time the old state will be restored unless that state was expired (fixed at 7 days I believe). Other option would be to always have it save the state irrespective if restore is set to True or False but only restore when restore is set to True.

Reply:
So I Mae it part of TemplateEntity and TriggerEntity now, overloading async_internal_added_to_hass so if state is not to be saved I then remove it accordingly.
This also fixes the issue if the user has restore enabled, disabled, and then enables it again. When entity loads on disabling the state is removed from last_states and thus will not be saved anymore. So when they then re-enable it would be as if it is the 1st time they enable it.